lib/ui/window.dart

```diff
@@ -145,16 +147,56 @@ class Locale {
   /// The primary language subtag must not be null. The region subtag is
   /// optional.
   ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null), assert(_languageCode != '');
+  /// The subtag values are _case sensitive_ and should be one of the valid
+  /// subtags according to CLDR supplemental data:
+  /// [language](http://unicode.org/cldr/latest/common/validity/language.xml),
+  /// [region](http://unicode.org/cldr/latest/common/validity/region.xml). The
+  /// primary language subtag should be two to three lowercase letters. The
+  /// region region subtag should be two uppercase letters or three digits. See
+  /// the [Unicode Language
+  /// Identifier](https://www.unicode.org/reports/tr35/#Unicode_language_identifier)
+  /// specification.
+  ///
+  /// Validity is not checked by default, but some methods may throw away
+  /// invalid data.
+  ///
+  /// See also:
+  ///
+  ///  * [new Locale.fromSubtags].
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       assert(_languageCode != ''),
+       scriptCode = null;
+
+  /// Creates a new Locale object.
+  ///
+  /// The keyword arguments specify the subtags of the Locale.
+  ///
+  /// The subtag values are _case sensitive_ and should be valid subtags
+  /// according to CLDR supplemental data:
+  /// [language](http://unicode.org/cldr/latest/common/validity/language.xml),
+  /// [script](http://unicode.org/cldr/latest/common/validity/script.xml),
+  /// [region](http://unicode.org/cldr/latest/common/validity/region.xml).
+  /// for each of languageCode, scriptCode and countryCode respectively.
+  ///
+  /// Validity is not checked by default, but some methods may throw away
+  /// invalid data.
+  const Locale.fromSubtags({
+    String languageCode = 'und',
+    this.scriptCode,
+    String countryCode,
+  }) : assert(languageCode != null),
+       assert(languageCode != ''),
+       _languageCode = languageCode,
+       assert(scriptCode != ''),
+       assert(countryCode != ''),
```

Comment:
let's add this assert to the other constructor too while we're at it

Reply:
With this I then manage to bring us back to the unit test that I disliked:

expect(const Locale('en'), isNot(new Locale('en', '')));

The existing Locale constructor is being tested to support both `null` and `''`, and furthermore being tested to treat these differently. Mark mentioned to me yesterday "in other languages it's been found most convenient to return empty strings for omitted subtags rather than null". I suggest Dart is quite null-happy and thus might not have the same concerns. Either way, that data point and the idea to forbid `''` both speak in favour of, at the very least, not treating `''` and `null` differently?

So: in-scope or out of scope for this Pull Request to change existing behaviour? I'm happy to consider this via a follow-up change too...

Reply:
It's fine to remove support for `''` as a value here IMHO (i.e. fix the test to verify that the empty string throws). No valid use of this API could have done that before. Having multiple ways to express the same value seems like a likely source of bugs (e.g. they'd have different hash codes presumably, unless we went out of our way to have them have the same hash code, at which point we're just wasting cycles on something that handled essentially for free by just banning the empty string).

`null` meaning "absent" is idiomatic in Dart. The empty string in Dart doesn't represent absence, it represents presence with a particular value (which happens to be the empty string, but could as easily be any other string).

Reply:
I've added the assert and removed it from the unit test, though I'm not sure how to actually trigger a failure. Looks like asserts are dropped.

From an attempt to add --enable-asserts to the dart invocation for locale_test.dart:

```
+ run_test flutter/testing/dart/locale_test.dart
+ out/host_debug_unopt/dart --enable-asserts out/host_debug_unopt/gen/frontend_server.dart.snapshot --sdk-root out/host_debug_unopt/flutter_patched_sdk --incremental --strong --target=flutter --packages flutter/testing/dart/.packages --output-dill out/host_debug_unopt/engine_test.dill flutter/testing/dart/locale_test.dart
Snapshot not compatible with the current VM configuration: the snapshot requires 'release strong no-type_checks no-asserts no-error_on_bad_type sync_async reify_generic_functions use_field_guards use_osr x64-sysv' but the VM has 'release strong no-type_checks asserts no-error_on_bad_type sync_async reify_generic_functions use_field_guards use_osr x64-sysv'
Isolate creation failed
```

I'd need to look into how one might tweak the snapshot later.